### ISSUE TITLE
Array return types are not typesafe

### ISSUE THREAD

**Issue by CEikermann** [Microsoft/TypeScript#6449]
Hey,

recently I found out that the compile does not check the types of classes in a array. I have created a example which should generate a compile error, but currently it does not.

http://www.typescriptlang.org/Playground#src=class%20BaseClass%20%7B%0A%09%0A%7D%0A%0Aclass%20ExtendedClass%20extends%20BaseClass%20%7B%0A%09%0A%7D%0A%0Aclass%20NotExtendedClass%20%7B%0A%09%0A%7D%0A%0Aclass%20Test%20%7B%0A%09%0A%09testFunc()%3A%20Array%3CBaseClass%3E%20%7B%0A%20%20%20%20%20%20%20%20return%20%5B%0A%09%09%09new%20ExtendedClass()%2C%0A%09%09%09new%20NotExtendedClass()%2C%0A%09%09%5D%3B%0A%20%20%20%20%7D%0A%09%0A%7D

TypeScript Version: 1.7.5

Best regards,
Christian

**Comment by aleksey-bykov** [comment]
TypeScript uses structural typing, it means that any 2 types are considered the same as long as their shape is the same. Shape is a set of properties with their types.


The shape of the base class (no properties) is the same as the shape of the derived class (no properties) is the same as the shape of unrelated class (no properties).

**Comment by CEikermann** [comment]
Thank you for the explanation.

That is totally correct, after I have added a private property to the base class the compile detects the invalid usage of NotExtendedClass class.

http://www.typescriptlang.org/Playground#src=class%20BaseClass%20%7B%0A%09%0A%7D%0A%0Aclass%20ExtendedClass%20extends%20BaseClass%20%7B%0A%09%0A%7D%0A%0Aclass%20NotExtendedClass%20%7B%0A%09%0A%7D%0A%0Aclass%20Test%20%7B%0A%09%0A%09testFunc()%3A%20Array%3CBaseClass%3E%20%7B%0A%20%20%20%20%20%20%20%20return%20%5B%0A%09%09%09new%20ExtendedClass()%2C%0A%09%09%09new%20NotExtendedClass()%2C%0A%09%09%5D%3B%0A%20%20%20%20%7D%0A%09%0A%7D

To the repo collaborators: should I leave this ticket open due to the fact the compile does not recognize the invalid usage as long the base class is empty ? But I think, it is currently just not possible

**Comment by sandersn** [comment]
No, this is by design. `BaseClass` is treated as an alias for `{}`, which is the object that has no properties at all. It's technically the base class of all objects. If you try to use the array you'll see that only properties in `BaseClass` are visible. It's easier to tell [by adding a commonly used property name like `name`](http://www.typescriptlang.org/Playground#src=class%20BaseClass%20%7B%0A%09%0A%7D%0A%0Aclass%20ExtendedClass%20extends%20BaseClass%20%7B%0A%09%0A%7D%0A%0Aclass%20NotExtendedClass%20%7B%0A%09%0A%7D%0A%0Aclass%20Test%20%7B%0A%09%0A%09testFunc()%3A%20Array%3CBaseClass%3E%20%7B%0A%20%20%20%20%20%20%20%20return%20%5B%0A%09%09%09new%20ExtendedClass()%2C%0A%09%09%09new%20NotExtendedClass()%2C%0A%09%09%5D%3B%0A%20%20%20%20%7D%0A%09%0A%7D).

Notice that `a[0].name` is legal but `a[0].nope` is not.


```ts
class BaseClass {
	name: string
}
class ExtendedClass extends BaseClass {	
}
class NotExtendedClass {
	name: string	
}
class Test {	
	testFunc(): Array<BaseClass> {
        return [
			new ExtendedClass(),
			new NotExtendedClass(),
		];
    }
}
let a = new Test().testFunc();
a[0].name + a[0].nope;
```
